Re: Add an API for Syncing Drafts Across the Zulip Server

Hypro999 opened the pull request:
Check out: https://github.com/zulip/zulip/pull/15880 for more info.

Hypro999 commented:
I'll resolve the conflicts when it looks like this PR will be reviewed.

Hypro999 commented:
@timabbott: Rebased and ready for review.

Hypro999 commented:
The `Zulip CI / Ubuntu 20.04 Focal (Python 3.8, backend) (pull_request)` test failure is irrelevant:
```
Testing only internal links in documentation...
2021-05-12 04:47:41 [scrapy.downloadermiddlewares.retry] ERROR: Gave up retrying <HEAD https://zulip.com/help/add-custom-profile-fields> (failed 3 times): 502 Bad Gateway
2021-05-12 04:47:41 [scrapy.downloadermiddlewares.retry] ERROR: Gave up retrying <HEAD https://zulip.com/help/restrict-name-and-email-changes> (failed 3 times): 502 Bad Gateway
2021-05-12 04:47:41 [scrapy.downloadermiddlewares.retry] ERROR: Gave up retrying <HEAD https://zulip.com/help/invite-new-users> (failed 3 times): 502 Bad Gateway
2021-05-12 04:47:41 [documentation_crawler] ERROR: Please check link: https://zulip.com/help/add-custom-profile-fields
2021-05-12 04:47:41 [documentation_crawler] ERROR: Please check link: https://zulip.com/help/restrict-name-and-email-changes
2021-05-12 04:47:42 [documentation_crawler] ERROR: Please check link: https://zulip.com/help/invite-new-users
Failed!
```

@timabbott: Rebased and ready for review.

timabbott commented on zerver/models.py:
```diff
@@ -1520,6 +1520,9 @@ class UserProfile(AbstractBaseUser, PermissionsMixin, UserBaseSettings):
     EMAIL_REMINDER_WAITPERIOD = 24
 
+    # Whether or not the user wants to sync their drafts.
+    enable_drafts_synchronization = models.BooleanField(default=True)
```
I think this should be moved to `UserBaseSettings`.

timabbott commented on zerver/views/user_settings.py:
```diff
@@ -271,6 +272,12 @@ def json_change_settings(
+        user_profile.enable_drafts_synchronization = enable_drafts_syncing
+        user_profile.save(update_fields=["enable_drafts_synchronization"])
+        if enable_drafts_syncing is False:
+            Draft.objects.filter(user_profile=user_profile).delete()
```
I think we want the API name and database name to match; so `enable_drafts_synchronization != enable_drafts_syncing` is a bug.  I'd probably do `enable_drafts_synchronization` as the one to keep.

timabbott replied:
We can likely also use the `property_types` system for this property; then you'd just need to move the `Draft.objects.filter` line to the `do_change_display_setting`  function in `actions.py`.

Hypro999 replied:
All of the properties in `UserBaseSettings.property_types` seem to be display related properties, so I'm against putting this there. Are you sure you want to move it?

timabbott replied:
Yes, see https://chat.zulip.org/#narrow/stream/378-api-design/topic/settings.20endpoints/near/1237310.  Basically, (1) `display_settings` had become the catch-all place for user settings that were not notification settings and (2) we're partway through merging these into a single "settings" object, so that the API doesn't depend on this sort of detail.

Hypro999 replied:
I couldn't find any method named `do_change_display_setting`, did you mean `do_set_user_display_setting`?

timabbott commented:
I posted a batch of comments that should let us merge the first few commits.  Thanks for rebasing this @Hypro999!

Hypro999 commented:
@timabbott thanks for the review. I'm ready for another round.

timabbott commented:
I merged the first two commits as the series ending with b5cd232a2ea946d0ca80802ac833b0dbe8db1aa6, after renumbering migrations (due to another PR I merged earlier today) and changing the error message string.

Hypro999 commented:
@timabbott, I've gone ahead and moved `enable_drafts_synchronization` to `property_types`. Can you take another look at this now?

What's changed:
1. I've split out the (currently) first commit into two parts after I accidentally merged them together in the last round of review.
2. Rebased against master.
3. Added a new commit at the end to make `enable_drafts_synchronization` a `property_type` and do all of the refactoring involved.

timabbott commented:
I rebased this and reorganized what remains into 3 commits that each do one thing, which makes it a lot more readable.  I also did numerous minor edits to the API documentation, in particular around documenting the feature levels clearly and changes that simplify the documentation of API endpoints (e.g. we no longer need template files for these).

There's a few things that are outstanding for this backend:
* @Hypro999 it'd be great if you can rereview the version I pushed back here for correctness.
* For the "Delete a draft" endpoint, should we just send the `draft_id` in the event?  I think that'd be cleaner, since otherwise you're getting a blob with a bunch of values that are "what was just deleted".  I think `draft_id: 14` is the right format (i.e. making sure the field name makes clear what it does).
* I don't understand the limit of at most 10 drafts. I think we should change that to 1000 and also make it a constant `MAX_DRAFTS_IN_REGISTER_RESPONSE`, so that we can have the test override that constant and make it 5 to verify the logic for having too many. I would not expect even 1000 drafts to be a problem performance-wise.
* We should make sure CI passes after all the fiddling I did.  The items above here are the only blockers; @Hypro999 let me know when they are done, and I can merge this PR (and we can open follow-up issues for what remains).
* The API documentation doesn't have Python examples, which we can do with `call_endpoint`.
* The API documentation for fetching drafts displays the documentation reasonably, but for the endpoints where you're submitting a set of draft objects, the API docs don't actually display the format in any detail.  @MSurfer20 any ideas what's up here?
![image](https://user-images.githubusercontent.com/2746074/127412784-e459838f-4879-45a7-8b28-76a2a2cf96ec.png)

Hypro999 commented:
> For the "Delete a draft" endpoint, should we just send the draft_id in the event? I think that'd be cleaner...

Sure thing! I agree that maybe sending the entire draft might be a bit too bloated. Plus it's probably redundant for most use cases.

> I don't understand the limit of at most 10 drafts. I think we should change that to 1000 and also make it a constant MAX_DRAFTS_IN_REGISTER_RESPONSE

I sort of just chose that arbitrarily. My mistake, guess I should have avoided the magic number 😅. I'll change this.

timabbott commented:
Merged, after a few changes:
* Split the `populate_db` stuff to its own commit.
* Edited the copy for some of the API endpoints
* Removed stale documentation of an `update_draft_settings` event type that does not exist that you appear to have added since yesterday.

Huge thanks for doing this @Hypro999!

timabbott commented:
Do you want to do the frontend PR for this as well?  I imagine we'll test-deploy it on chat.zulip.org once it's mostly working.

Hypro999 commented:
> Do you want to do the frontend PR for this as well?  I imagine we'll test-deploy it on chat.zulip.org once it's mostly working.

I can't guarantee that I'll be able to commit the time for it. So if anyone else wants to do this they're welcome. However, as of now, I do kind of want to play around with the frontend implementation a bit. So even if I don't end up writing the implementation, I'll help out with the reviews.
